curp/src/server/raw_curp/mod.rs

```diff
@@ -0,0 +1,806 @@
+        st.randomize_timeout_ticks(); // regenerate timeout ticks
+        self.reset_election_tick();
+        self.ctx.hb_opt.store(false, Ordering::Relaxed);
+        debug!(
+            "{} updates to term {term} and becomes a follower",
+            self.id()
+        );
+    }
+
+    /// Reset election tick
+    fn reset_election_tick(&self) {
+        self.ctx.election_tick.store(0, Ordering::Relaxed);
+    }
+
+    /// Check whether `commit_index` can be updated to i
+    fn can_update_commit_index_to(
+        &self,
+        lst: &LeaderState,
+        log: &Log<C>,
+        i: usize,
+        cur_term: u64,
+    ) -> bool {
+        if log.commit_index >= i {
+            return false;
+        }
+
+        // don't commit log from previous term
+        if log.entries[i].term() != cur_term {
+            return false;
+        }
+
+        let replicated_cnt: u64 = self
+            .ctx
+            .others
+            .iter()
+            .filter(|&id| lst.match_index[id] >= i)
+            .count()
+            .numeric_cast();
+        replicated_cnt + 1 >= self.quorum()
+    }
+
+    /// Recover from all voter's spec pools
+    fn recover_from_spec_pools(
+        &self,
+        st: &mut State,
+        log: &mut Log<C>,
+        spec_pools: &HashMap<ServerId, Vec<Arc<C>>>,
+    ) {
+        debug!("{} collected spec pools: {spec_pools:?}", self.id());
+
+        let mut cmd_cnt: HashMap<ProposeId, (Arc<C>, u64)> = HashMap::new();
+        for cmd in spec_pools.values().flatten().cloned() {
+            let entry = cmd_cnt.entry(cmd.id().clone()).or_insert((cmd, 0));
+            entry.1 += 1;
+        }
+
+        // get all possibly executed(fast path) commands
+        let existing_log_ids: HashSet<ProposeId> = log
+            .entries
+            .iter()
+            .flat_map(|entries| entries.cmds().iter().map(|cmd| cmd.id()).cloned())
+            .collect();
+        let recovered_cmds = cmd_cnt
+            .into_values()
+            // only cmds whose cnt >= 3/4 can be recovered
+            .filter_map(|(cmd, cnt)| (cnt >= self.superquorum()).then(|| cmd))
+            // dedup in current logs
+            .filter(|cmd| {
+                // TODO: better dedup mechanism
+                !existing_log_ids.contains(cmd.id())
+            })
+            .collect_vec();
+
+        let mut cb_w = self.ctx.cb.write();
+        let mut sp_l = self.ctx.sp.lock();
+
+        let term = st.term;
+        for cmd in recovered_cmds {
+            debug!(
+                "{} recovers speculatively executed cmd {:?} in log[{}]",
+                self.id(),
+                cmd.id(),
+                log.entries.len(),
+            );
+            let _ig_sync = cb_w.sync.insert(cmd.id().clone()); // may have been inserted before
+            let _ig_spec = sp_l.insert(Arc::clone(&cmd)); // may have been inserted before
+            log.entries.push(LogEntry::new(term, &[cmd]));
```

Comment:
I think it's the same here, I inserted the cmd to speculative pool, push the entry into log and the log will soon be replicated on each nodes(because on next tick, the server will find inconsistencies and start calibrating and the recovered cmd will be replicated on others).

Reply:
Calibrating does almost the same work as AE with a start latency. If we can tolerant this latency, I think it's fine.

Reply:
I will create an issue to track this optimization.